runtime/src/leader_bank_notifier.rs

```diff
@@ -0,0 +1,161 @@
+use {
+    crate::bank::Bank,
+    solana_sdk::slot_history::Slot,
+    std::{
+        sync::{Arc, Condvar, Mutex, Weak},
+        time::{Duration, Instant},
+    },
+};
+
+/// Tracks leader status of the validator node and notifies when:
+///     1. A leader bank initiates (=PoH-initiated)
+///     2. A leader slot completes (=PoH-completed)
+#[derive(Debug, Default)]
+pub struct LeaderBankNotifier {
+    /// Current state (slot, bank, and status) of the system
+    state: Mutex<SlotAndBankWithStatus>,
+    /// CondVar to notify status changes and waiting
+    condvar: Condvar,
+}
+
+/// Leader status state machine for the validator.
+#[derive(Copy, Clone, Debug, Default, PartialEq, Eq)]
+enum Status {
+    /// The leader bank is not currently available. Either not initialized, or PoH-completed bank.
+    #[default]
+    StandBy,
+    /// PoH-initiated bank is available.
+    InProgress,
+}
+
+#[derive(Debug, Default)]
+struct SlotAndBankWithStatus {
+    status: Status,
+    slot: Slot,
+    bank: Weak<Bank>,
+}
+
+impl LeaderBankNotifier {
+    /// Set the status to `InProgress` and notify any waiting threads
+    /// if the status was not already `InProgress`.
+    pub fn set_in_progress(&self, bank: &Arc<Bank>) {
+        let mut state = self.state.lock().unwrap();
+        if matches!(state.status, Status::InProgress) {
```

Comment:
i'd add `assert_eq!(state.slot, bank.slot());` just inside at the top of this if body.

moreover, is it necessary to make this fn idempotent? it's only called only once from replay_stage, which should guarantee idempotent isn't needed.

so, I'd remove remove this `if` altogether and replace it with `assert_matches!(state.status, Status::StandBy);` in that way, human reasoning can be simplified and serious logic error at caller side will be caught. imo, idempotent fns are handy sometimes, but it's a kind of variant to swallowing errors (obviously code smell).

Reply:
removed the if altogether: 8c1b5e2e584fc91e40a1fc770a224823d7499602